src/qutip_cupy/dense_functions.py

```diff
@@ -29,3 +29,18 @@ def trace_cupydense(cp_arr):
     # @TODO: whnen qutip allows it we should remove this call to item()
     # as it takes a time penalty commmunicating data from GPU to CPU.
     return cp.trace(cp_arr._cp).item()
+
+
+def project_dense(state):
+    """
+    Calculate the projection |state><state|.  The shape of `state` will be used
+    to determine if it has been supplied as a ket or a bra.  The result of this
+    function will be identical is passed `state` or `adjoint(state)`.
+    """
+
+    if state.shape[1] == 1:
+        return CuPyDense._raw_cupy_constructor(cp.outer(state._cp, state.adjoint()._cp))
+    elif state.shape[0] == 1:
+        return CuPyDense._raw_cupy_constructor(cp.outer(state.adjoint()._cp, state._cp))
+    else:
+        raise TypeError("state must be a ket or a bra.")
```

Comment:
```suggestion
        raise ValueError("state must be a ket or a bra.")
```
I think this is technically a `ValueError` since the type is correct.